deepeval/evaluate.py

```diff
@@ -69,31 +71,96 @@ def create_api_test_case(
     )
 
 
+def generate_cache_key(test_case: LLMTestCase):
+        # Explicitly handle None values for context and retrieval_context
+        context = sorted(test_case.context) if test_case.context else []
+        retrieval_context = sorted(test_case.retrieval_context) if test_case.retrieval_context else []
+
+        # Create a dictionary with the relevant fields
+        cache_data = {
+            "input": test_case.input,
+            "expected_output": test_case.expected_output,
+            "context": context,
+            "retrieval_context": retrieval_context,
+        }
+        
+        # Sort the dictionary by key to ensure consistent ordering
+        sorted_cache_data = dict(sorted(cache_data.items()))
+        
+        # Convert the sorted dictionary to a JSON string to use as a cache key
+        # JSON serialization ensures that the data structure is converted to a string format that can be hashed
+        cache_key = json.dumps(sorted_cache_data)
+        
+        return cache_key 
+
 def execute_test_cases(
     test_cases: List[LLMTestCase],
     metrics: List[BaseMetric],
     save_to_disk: bool = False,
 ) -> List[TestResult]:
+    
     test_results: List[TestResult] = []
     test_run_manager.save_to_disk = save_to_disk
+
     for index, test_case in enumerate(test_cases):
+
+        test_run = test_run_manager.get_test_run()
+        cached_test_run = test_run_cache_manager.get_cached_test_run()
+        key = generate_cache_key(test_case=test_case)
+        look_up_map = cached_test_run.test_cases_lookup_map
+
+        #########################################
+        # Get cached data if available
+        ##########################################
+        already_in_cache = (key in look_up_map)
+        hyperparameter_same = cached_test_run.hyperparameters == test_run.hyperparameters
+        actual_output_exists = False
+        metrics_metadata_exists = False
+        cached_api_test_case = None
+
+        if already_in_cache:
+            cached_api_test_case = look_up_map[key]
+            actual_output_exists = cached_api_test_case.actual_output != None
+            if actual_output_exists:
+                metrics_metadata_exists = cached_api_test_case.metrics_metadata != None
+        can_use_cache = (already_in_cache and hyperparameter_same and actual_output_exists and metrics_metadata_exists)
+
+        cached_metrics_metadata = {} 
+        if can_use_cache:
+            metrics_metadata = look_up_map[key].metrics_metadata
+            cached_metrics_metadata = {metadata.metric: metadata for metadata in metrics_metadata}
```

Comment:
naming a bit confusing, the metrics metadata from lookup map is already the cached metrics metadata (make it cached metrics metadata map maybe?)